gc_derive/src/lib.rs

```diff
@@ -78,3 +97,23 @@ decl_derive!([Finalize] => derive_finalize);
 fn derive_finalize(s: Structure<'_>) -> proc_macro2::TokenStream {
     s.unbound_impl(quote!(::gc::Finalize), quote!())
 }
+
+decl_derive!([EmptyTrace] => derive_empty_trace);
+
+// TODO: Does not work on self-referential types
+fn derive_empty_trace(mut s: Structure<'_>) -> proc_macro2::TokenStream {
+    // Add where bounds for all bindings manually because synstructure only adds them if they depend on one of the parameters.
+    let mut where_predicates = Vec::new();
+    for v in s.variants() {
+        for bi in v.bindings() {
+            let ty = &bi.ast().ty;
+            let span = ty.span();
+            where_predicates.push(parse_quote_spanned! { span=> #ty: ::gc::EmptyTrace });
+        }
+    }
+    for p in where_predicates {
+        s.add_where_predicate(p);
+    }
+    s.add_bounds(AddBounds::None);
```

Comment:
Isn’t this what `AddBounds::Fields` is for?

```suggestion
    s.add_bounds(AddBounds::Fields);
```

Reply:
It's what I tried to explain in the comment at the top. If you use `AddBounds::Fields` then it skips any field that does not mention a parameter. But we need those to implement `EmptyTrace` too.

Reply:
Ah, yeah, I suggest generating assertions for them; the typical way that custom derives work is that the actual method code enforces that each field implements the trait. This trait doesn't have a method (we could add a hidden defaulted one). 

Otherwise the field trait bounds escape into the docs.